src/control/lib/control/server_meta.go

```diff
@@ -0,0 +1,192 @@
+//
+// (C) Copyright 2020 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package control
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+
+	"github.com/daos-stack/daos/src/control/common/proto/convert"
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+	"github.com/daos-stack/daos/src/control/server/storage"
+	"github.com/daos-stack/daos/src/control/system"
+)
+
+type (
+	// SmdDevice contains DAOS storage device information, including
+	// health details if requested.
+	SmdDevice struct {
+		UUID      string                    `json:"uuid"`
+		TargetIDs []int32                   `hash:"set" json:"tgt_ids"`
+		State     string                    `json:"state"`
+		Rank      system.Rank               `json:"rank"`
+		Health    *storage.NvmeDeviceHealth `json:"health"`
+	}
+
+	// SmdPool contains the per-server components of a DAOS pool.
+	SmdPool struct {
+		UUID      string      `json:"uuid"`
+		TargetIDs []int32     `hash:"set" json:"tgt_ids"`
+		Blobs     []uint64    `hash:"set" json:"blobs"`
+		Rank      system.Rank `hash:"set" json:"rank"`
+	}
+
+	// SmdPoolMap provides a map from pool UUIDs to per-rank pool info.
+	SmdPoolMap map[string][]*SmdPool
+
+	// SmdInfo encapsulates SMD-specific information.
+	SmdInfo struct {
+		Devices []*SmdDevice `hash:"set" json:"devices"`
+		Pools   SmdPoolMap   `json:"pools"`
+	}
+
+	// SmdQueryReq contains the request parameters for a SMD query
+	// operation.
+	SmdQueryReq struct {
+		unaryRequest
+		OmitDevices      bool
+		OmitPools        bool
+		IncludeBioHealth bool
+		SetFaulty        bool
```

Comment:
Seems a little weird to me to have an action like injecting a fault as part of the query request. I would think that should be its own API entry point.

Reply:
Yeah, I considered that. This whole family of functionality is a little weird though, and my goal with this PR was to try to make it a bit more coherent from the API user's perspective. I'd like to see how this stuff evolves before pulling it apart again. If we get more fault injection features then I agree that it would make sense to throw this into that basket.